### ISSUE TITLE
Runtime issue when using .net standard dll in .net framework project

### ISSUE THREAD

**Issue by OldWarrior3000** [dotnet/runtime#55809]
Hi,

we are currently experiencing an issue when we want to use a .net standard DLL into our .net framework application.

The library are we are using is a client DLL which runs and HTTP  request on a micro service and deserializes the response from JSON into an object.

During the JSON deserialisation process (call to System.Text.Json) we are getting the following error:

```
Could not load file or assembly 'System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
```
We already tried to follow the suggestions from this issue [https://github.com/dotnet/runtime/issues/26370](https://github.com/dotnet/runtime/issues/26370)

Here all detailed list of all things we tried but without any success:
- Use the nuget package with .net standard 2.0
- Reference the project directly (framework target --> .net standard 2.0)
- Reference the project directly (framework target --> .net framework 4.8)
- Adding binding redirects as suggested in issue #26370
-  Adding AutogeneratedBindingRedirects and GenerateBindingRedirectsOutputType into the csproj
```xml
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
```
- Removed and re-installed the dependent packages System.Memory and System.Text.Json 
- Recompilation and restart of IIS

However, this issue only occurs when invoking the client from the application itself. When the search is invoked from a test project, (in the same solution) it works correctly and returns results without any problems.

The application in which we are trying to integrate the library contains a lot of reference to other Nuget packages and DLL's this could be the possible cause of the problem but we are not sure how to isolate the issue.

All help is very much appreciated. Thanks in advance

Environment:
- Main project: .net framework 4.8, web project with ASP.net and EntityFramework
- Client project: .net standard 2.0 (tried with multi target for framework 4.8)

**Comment by jeffhandley** [comment]
Thanks for the detailed report, @OldWarrior3000. This will require some investigation which we won't likely be able to complete very soon. For the time being, I'm going to mark this as `Future` and `needs further triage`. Please let us know if you get unblocked in the meantime.

**Comment by danmoseley** [comment]
Perhaps @joperezr is familiar with this.

**Comment by joperezr** [comment]
Yup, these kind of problems are right down my alley so I’m happy to help investigate. In most cases, problem is down to the configuration pf the project file so there is no “one solution” that fixes all, but with some investigation we should be able to figure out what is going on and how to fix it. 
The quickest and easiest way to investigate is if you are able to come up with a small repro project that you can share. If you can do that, I can pull that project and figure out the fix. 

If you can’t come up with a repro project, the next best thing is a build binlog of your project in order to figure out how are you referencing the library and what is happening at build time. In order to produce a binlog, run the following command in a developer command prompt: `msbuild yourAspNetProject.csproj /t:rebuild /bl` and that will generate a msbuild.binlog that I will ask you to share in order to investigate.

Let me know if you have any problems trying to provide any of the two things I asked above.

**Comment by joperezr** [comment]
FWIW, my hunch here is that your aspnet project is using packages.config, and the netstandard library might have a dependency on System.Numerocs.Vectors package and that transitive reference might not be flowing over to asp.net, which would explain why unit tests are fine, but aspnet project isn’t. If this is the case, fix can be either to migrate the project to packagereference (so that dependencies flow automatically) or to manually add a reference to system.numerics.vectors package from your asp.net project.

**Comment by OldWarrior3000** [comment]
Sorry, I cannot share solution as it is a propitiatory monolith and we managed to fix the issue in the meanwhile.

I think that you're right. After we posted this issue, we reverted all of the changes on the solution and then reinstalled/updated all of the NuGet packages that we needed to run the solution. 

After each NuGet installation/update, we attempted to run both the tests and the user action that triggered this issue to see if it would happen again. The problem started happening again when we installed a NuGet that has a reference to System.Text.Json (which has a dependency on System.Numerics.Vectors).

Since this solution has multiple projects, we had a hunch that there was a problem with versions of the reference to System.Text.Json on different projects. For some reason, consolidate didn't fix that issue. We then went manually through each packages.config file to check if the versions were the same everywhere (they weren't), and then updated, or even removed the entire dependency tree and reinstall the NuGets again. With that, we managed to get the solution working.

In my opinion (but this is just a guess), the problem occurred because the process of installing/updating/consolidating the NuGet packages didn't correctly adjust the packages.config files in the entire solution.
   
Thanks for your help! You can close the issue.

**Comment by Joe4evr** [comment]
> We then went manually through each packages.config file to check if the versions were the same everywhere (they weren't)

🍝 Just an idea, but if you can migrate declaring your dependencies with `PackageReference`, you *could* make the versions easier to synchronize by using a .props file:
```xml
<!-- Defined in some .props file that can get automatically imported -->
<SysJsonVersion>5.0.2</SysJsonVersion>

<!-- Any project that depends on S.T.J -->
<PackageReference Include="System.Text.Json" Version="$(SysJsonVersion)" />
```
Now there's only one place you need to adjust to update a dependency for all the projects.

**Comment by OldWarrior3000** [comment]
Thanks @Joe4evr, that's a great advice!

**Comment by tannergooding** [comment]
> Now there's only one place you need to adjust to update a dependency for all the projects.

Even easier than that is to use a `.targets` file.

In your project file:
```xml
<PackageReference Include="System.Text.Json" />
```

In a `Directory.Build.targets` in your repo root (or correspondingly correct location):
```xml
<PackageReference Update="System.Text.Json" Version="5.0.2" />
```

For example:
* csproj: https://github.com/terrafx/terrafx/blob/main/sources/Core/TerraFX.csproj#L10
* targets: https://github.com/terrafx/terrafx/blob/main/Directory.Build.targets#L35-L48

And an example of a shared `PackageReference` in a `Directory.Build.props`:
* References for all source projects: https://github.com/terrafx/terrafx/blob/main/sources/Directory.Build.props#L26
* References for all test projects: https://github.com/terrafx/terrafx/blob/main/tests/Directory.Build.props#L25-L29

etc
